profile-widget/diveprofileitem.cpp

```diff
@@ -161,37 +166,89 @@ void DiveProfileItem::replot(const dive *d, int from, int to, bool in_planner)
+	const plot_data *data = dataModel.data().entry;
+	const int max_peaks = (data[to].sec - data[from].sec) / half_interval + 1;
+	struct Peak {
+		int range_from;
+		int range_to;
+		int peak;
+	};
+	std::vector<Peak> peaks;
+	peaks.reserve(max_peaks);
+	int highest_peak = std::max_element(data + from, data + to, comp_depth) - data;
+	if (data[highest_peak].depth < min_depth)
+		return;
+	peaks.push_back(Peak{ from, to, highest_peak });
+	for (int act = 0; act < (int)peaks.size(); ++act) {
+		plot_depth_sample(data[peaks[act].peak], Qt::AlignHCenter | Qt::AlignTop, getColor(SAMPLE_DEEP));
+
+		// Clear to the left and right of peak and add new peaks if there is place
+		const Peak &act_peak = peaks[act];
+		const plot_data &act_sample = data[act_peak.peak];
+		int valley = act_peak.peak;
+
+		// Search for first sample outside minimum range to the right.
+		int new_from;
+		for (new_from = act_peak.peak + 1; new_from + 3 < act_peak.range_to; ++new_from) {
+			if (data[new_from].sec > act_sample.sec + half_interval)
+				break;
+			if (data[new_from].depth < data[valley].depth)
+				valley = new_from;
 		}
-
-		if ((entry == min_entry) && entry->depth / 100 != last) {
-			plot_depth_sample(entry, Qt::AlignHCenter | Qt::AlignTop, getColor(SAMPLE_SHALLOW));
-			last = entry->depth / 100;
+		// Continue search until peaks reach the minimum prominence (height from valley).
+		for ( ; new_from + 3 < act_peak.range_to; ++new_from) {
+			if (data[new_from].depth >= data[valley].depth + min_prominence) {
+				int new_peak = std::max_element(data + new_from, data + act_peak.range_to, comp_depth) - data;
+				if (data[new_peak].depth < min_depth)
+					break;
+				peaks.push_back(Peak{ new_from, act_peak.range_to, new_peak });
+
+				plot_depth_sample(data[valley], Qt::AlignHCenter | Qt::AlignBottom, getColor(SAMPLE_SHALLOW));
+				break;
+			}
+			if (data[new_from].depth < data[valley].depth)
+				valley = new_from;
 		}
 
-		if (entry->depth != last)
-			last = -1;
+		valley = act_peak.peak;
+
+		// Search for first sample outside minimum range to the left.
+		int new_to;
+		for (new_to = act_peak.peak - 1; new_to >= act_peak.range_from + 3; --new_to) {
+			if (data[new_to].sec + half_interval < act_sample.sec)
+				break;
+			if (data[new_to].depth < data[valley].depth)
+				valley = new_to;
+		}
+		// Continue search until peaks reach the minimum prominence (height from valley).
+		for ( ; new_to >= act_peak.range_from + 3; --new_to) {
+			if (data[new_to].depth >= data[valley].depth + min_prominence) {
+				int new_peak = std::max_element(data + act_peak.range_from, data + new_to, comp_depth) - data;
+				if (data[new_peak].depth < min_depth)
+					break;
+				peaks.push_back(Peak{ act_peak.range_from, new_to, new_peak });
+
+				plot_depth_sample(data[valley], Qt::AlignHCenter | Qt::AlignBottom, getColor(SAMPLE_SHALLOW));
+				break;
+			}
+			if (data[new_to].depth < data[valley].depth)
+				valley = new_to;
+		}
```

Comment:
this code is so temptingly similar to what's above... but making it direction independent would completely break my brain, I'm sure...

Reply:
Indeed, I also thought that having the same code twice is probably less confusing.

One could of course do this _extremely_ elegantly as a template taking iterators where one passes forward- and backward-iterators, respectively. But I have a feeling that you wouldn't be happy about the resulting piece of art.